subsys/net/ip/net_if.c

```diff
@@ -3182,14 +3182,15 @@ static struct in6_addr *net_if_ipv6_get_best_match(struct net_if *iface,
 				continue;
 			}
 
-			if (len == *best_so_far &&
-			    ipv6->unicast[i].addr_state == NET_ADDR_DEPRECATED &&
+			if (ipv6->unicast[i].addr_state == NET_ADDR_DEPRECATED &&
```

Comment:
This addresses the case where deprecated address comes after preferred one, but there's another scenario in which the preferred address may fail the `len >= *best_so_far` condition and hence it may not be selected over the deprecated address, right?

Reply:
There is a check
```
	if (len >= *best_so_far) {
```
few lines above. It should select the best prefix, we then skip the deprecated address if the other one is preferred one. I tried to construct the tests so that they verify this. Please clarify more what you mean if I missed something.

Reply:
I think the order of processing addresses shouldn't matter.

In case the address array contains:
[0] Preferred address for which `get_diff_ipv6` returns 32
[1] Deprecated address for which `get_diff_ipv6` returns 64

the code will correctly choose [0] despite [1] being a better match prefix-wise.

However if the order of elements is opposite:
[0] Deprecated address for which `get_diff_ipv6` returns 64
[1] Preferred address for which `get_diff_ipv6` returns 32

the code will choose [0] because [1] will be discarded early due to `len >= *best_so_far` not being satisfied (`32 >= 64`). I think the code should define some kind of `is_address_better` predicate and this predicate should exit early whenever one address is preferred and another one is deprecated.

Reply:
I think @Damian-Nordic is right, this will fail if the deprecated address is checked first, for example:
1. We first check deprecated address, with 64-bit prefix match. `best_so_far` is set to 64
2. In the second iteration we check preferred address with 40-bit prefix match. The preferred address should take preference, but since the prefix match is worse than `best_so_far`, we won't even check if the current address is deprecated.

I think we need to check `addr_state` for deprecated address before we do the prefix-match check.

Reply:
Indeed, good point. I will rework the code.